sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel_arg.asciidoc

```diff
@@ -0,0 +1,147 @@
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+When launching kernels that are represented as `sycl::kernel` objects,
+developers must pass arguments via the `set_arg` or `set_args` functions. These
+functions accept trivially copyable types.
+
+There are situations where developers would like to pass a raw byte
+representation of the kernel argument to the backend (e.g., when a single
+`parallel_for` in the source code is used to invoke multiple kernels with
+different arguments types, or when passing an argument to a built-in kernel
+for which there is no equivalent type defined on the host).
+
+=== Usage example
+
+[source,c++]
+----
+int a;
+char* opaque_type;
+int nbytes;
+...
+h.set_arg(0, a);
+h.set_arg(1, sycl::ext::oneapi::raw_kernel_arg(opaque_type, nbytes));
+h.parallel_for(range, kernel);
+----
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_RAW_KERNEL_ARG` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== The `raw_kernel_arg` class
+
+This extension adds a new `raw_kernel_arg` class that can be used to declare
+kernel arguments via a raw byte representation.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+  class raw_kernel_arg {
+  public:
+    raw_kernel_arg(void* bytes, size_t count);
+  };
+
+} // namespace sycl::ext::oneapi
+----
+
+[source,c++]
+----
+raw_kernel_arg(void* bytes, size_t count);
+----
+_Preconditions_: `bytes` must point to an array of at least `count` bytes.
+
+_Effects_: Constructs a `raw_kernel_arg` representing the `count` bytes
+starting at the address specified by `bytes`.
+
+=== Using a raw kernel argument
+
+Instances of `raw_kernel_arg` are passed to kernels via the existing `set_arg`
+and `set_args` functions defined by the SYCL specification.
+
+This extension adds a new overload of `set_arg`, as defined below.
+
+[_Note:_ The parameter pack accepted by `set_args` has no constraints, and
+therefore requires no modification to support `raw_kernel_arg`. _{endnote}_]
+
+[source,c++]
+----
+void set_arg(int argIndex, sycl::ext::oneapi::raw_kernel_arg&& arg);
```

Comment:
Why do we use `&&` here?  Wouldn't it be better to pass `const raw_kernel_arg&`?

Reply:
I copied the existing definitions: `set_arg` accepts `T&&` and `set_args` accepts `Ts&&`.

I think it would be weird if `set_arg` accepted `const raw_kernel_arg&` but `set_args` accepted `raw_kernel_arg&&` via the parameter pack.  But honestly, I'm not sure I understand the difference between the two options well enough to know what the impact of this would be.

Reply:
I see.  Yes, that would be weird.

Since `set_arg` has no constraints on `T`, why do we need to add this overload at all?  Can't we just document in this extension that it is valid to pass `raw_kernel_arg` to `set_arg` and `set_args`?

Reply:
There's no real explanation of how `set_arg` works in the specification; all it says is that the backend is responsible for defining how it works.  In practice, though, the value passed in usually has the same type as the kernel parameter.  Things like accessors are a special case, although I'd expect the types would still match if calling a `sycl::kernel` compiled from SYCL source (using the online compiler).

I think `set_arg(raw_kernel_arg)` is a different special case, and its behavior is backend-independent.  The kernel doesn't need to be written to accept a `raw_kernel_arg` -- it's the backend's responsibility to accept the raw bytes and use those as the value of the kernel argument, irrespective of its real type.

Reply:
Oh, I see.

What I meant was that because: 1) the definition of `set_args` does not place any restrictions on `T`; and
2) the definition of `set_args` says that it acts "as if each argument in `args` was passed to `set_arg`",  adding a new overload of `set_arg` is sufficient to change the behavior of `set_args`. 

If the definition of `set_args` had said it was only available if `T` was trivially copyable or device copyable, or made some reference to the kernel argument rules, then I would have defined a new overload of `set_args`.

Does that make sense?  Perhaps it would be clearer if the note said something like I wrote above:
> [_Note_: Since the definition of `set_args` says that it acts "as if each argument in `args` was passed to `set_arg`",  adding a new overload of `set_arg` is sufficient to change the behavior of `set_args`. - _end note_]

Reply:
OK, makes sense.  That new "note" wording does seem better,